### ISSUE TITLE
Duplicated .protos in multi-project build

### ISSUE THREAD

**Issue by mrgoltstein** [google/protobuf-gradle-plugin#470]
Version 0.8.15 of the protobuf-gradle-plugin throws an error in Android projects with Gradle 7 when trying to run unit tests. This issue does not occur on the previous version 0.8.14, and also not on Gradle versions prior to 7. It also doesn't occur when building the debug or production variant of the app.

Gradle 7 is not yet final but has released a milestone version.

```
Execution failed for task ':mymodule:extractIncludeDebugUnitTestProto'.
> Entry myprotofile.proto is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.0-milestone-1/dsl/org.gradle.api.file.CopySpec.html#org.gradle.api.file.CopySpec:duplicatesStrategy for details.
```

Let me know if you need any more details. I'm also open to contributing if that helps, so let me know if you'd like that.

### Steps to reproduce:

Invoke `./gradlew testDebugAndroidTest` on a project with a proto file. 

`myprotofile.proto`
```
syntax = "proto3";

option java_package = "com.app.my";
option java_multiple_files = true;

message MyMessage {
  int32 id = 1;
}
```

`build.gradle.kts`
```kotlin
import com.google.protobuf.gradle.generateProtoTasks
import com.google.protobuf.gradle.protobuf
import com.google.protobuf.gradle.protoc

plugins {
    id("com.android.library")
    id("com.google.protobuf:0.8.15")
}

android {
    compileSdkVersion(1)
}

dependencies {
    implementation("com.google.protobuf:protobuf-javalite:3.14.0")
}

protobuf {
    protoc {
        artifact = "com.google.protobuf:protoc:3.14.0"
    }

    generateProtoTasks {
        all().forEach { task ->
            task.builtins {
                create("java") {
                    option("lite")
                }
            }
        }
    }
}
```

### Workaround

Just in case anyone is interested, the only workaround I could find is to disable the `extractInclude` task for test configurations. This worked for my particular usecase but may not for yours.

```
tasks.configureEach {
    if (name.matches(Regex("^extractInclude([A-Za-z]+)TestProto$"))) {
        enabled = false
    }
}
```

**Comment by voidzcy** [comment]
Can you provide a reproducible sample project? I tried with Gradle `7.0-milestone-1` with AGP `4.2.0-alpha10` but everything works fine. It might have something to do with your project structure.

**Comment by voidzcy** [comment]
Thanks for the sample project. I can reproduce what you are describing, will get back to you when I got time looking into it.

**Comment by gpouzerate** [comment]
Any news on this issue? Experiencing the same here and couldn't get the workaround to work

**Comment by tom-haines** [comment]
I have added a PR https://github.com/google/protobuf-gradle-plugin/pull/487 that fixes the issue, and has a unit test that confirms same.  The reason is that gradle 7+ moves from warning to fail if duplicates strategy is not defined https://github.com/gradle/gradle/pull/15759 .

**Comment by pluttrell** [comment]
Is there a work around until this is resolved?

**Comment by tom-haines** [comment]
@pluttrell this workaround expands on suggestion from @mrgoltstein  with extra disabled  task added via broader regex
```
tasks.configureEach {
  if (name.matches(Regex("^.+([A-Za-z]+)TestProto$"))) {
    enabled = false
  }
}
```

**Comment by voidzcy** [comment]
Should be fixed by #487.

**Comment by gpouzerate** [comment]
I couldn't get the workaround described above to work. I had to make it like this:

```
tasks.withType<ProcessResources> {
    duplicatesStrategy = DuplicatesStrategy.INCLUDE
}


```

**Comment by pluttrell** [comment]
> @pluttrell this workaround expands on suggestion from @mrgoltstein with extra disabled task added via broader regex
> 
> ```
> tasks.configureEach {
>   if (name.matches(Regex("^.+([A-Za-z]+)TestProto$"))) {
>     enabled = false
>   }
> }
> ```

@tom-haines Thanks for this. I couldn't however get it or the original one to compile. The issue appears to be with the last double-quote in the regex.

I was also wondering what the implications are of disabling task, whereas the PR sets the `duplicatesStrategy` to include.

**Comment by pluttrell** [comment]
> I couldn't get the workaround described above to work. I had to make it like this:
> 
> ```
> tasks.withType<ProcessResources> {
>     duplicatesStrategy = DuplicatesStrategy.INCLUDE
> }
> ```

@gpouzerate Is it possible that this is a Kotlin only solution? With the groovy DSL, I'm seeing errors the following errors based on where it's placed. But it's weird because Intellij suggests that `withType` exists.

```
Could not get unknown property 'withType' for task set of type org.gradle.api.internal.tasks.DefaultTaskContainer.
```
or
```
No such property: withType for class: com.google.protobuf.gradle.ProtobufConfigurator
```

**Comment by mgodave** [comment]
I'm running into the same issue as @pluttrell. Would it be possible to provide a groovy workaround as well?

**Comment by pluttrell** [comment]
I see the fix has been merged. Does anyone know when a release will be pushed?

**Comment by voidzcy** [comment]
I can do for a release some time later this week or early next week.

**Comment by ejona86** [comment]
As I mention in #487, I think we shouldn't use DuplicateStrategy.INCLUDE as that papers over issues. If a user is seeing this legitimately they need to fix their dependencies to avoid duplicates, as there's simple no other way to have a consistent build result.

That said, it seems at least some cases of this are caused by a bug in the plugin. In a multi-project build the resources directory and the jar are both being used which means you will see duplicate files. In the case of gRPC that was two files: `grpclb/build/resources/main/grpc/lb/v1/load_balancer.proto` and `alts/build/tmp/expandedArchives/grpc-grpclb-1.39.0-SNAPSHOT.jar_fd0392f42b3b7665203deb295b39a7ec/grpc/lb/v1/load_balancer.proto`. We will need to fix that bug before we revert #487.

I used this to find where the _first_ source of load_balancer.proto, as FAIL normally just tells you the second source:
```diff
diff --git a/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy b/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy
index fc65bba..11a53e8 100644
--- a/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy
+++ b/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy
@@ -113,9 +113,10 @@ abstract class ProtobufExtract extends DefaultTask {
       spec.includeEmptyDirs = false
       spec.from(allProtoFiles)
       spec.include('**/*.proto')
+      allProtoFiles.asFileTree.visit { println it.file }
       spec.into(destDir)
       // gradle 7+ requires a duplicate strategy to be explicitly defined
-      spec.duplicatesStrategy = DuplicatesStrategy.INCLUDE
+      spec.duplicatesStrategy = DuplicatesStrategy.FAIL
     }
   }
 
```

**Comment by mrgoltstein** [comment]
Just discovered that I forgot to close this issue. Thank you for your help!
